### ISSUE TITLE
Handle and decode/report modbus errors

### ISSUE THREAD

**Issue by cbrake** [simpleiot/simpleiot#306]
The Velocio PLC may be a good device to test with.

**Comment by floriaanpost** [comment]
Thank you for the good work on this library! 

Not sure if it is related, but I am currently trying the modbus library (standalone) to create a modbus (RTU) server and considering using it in production (the thing I like most is that it is created with and `io.ReadWriteCloser`, making it easy to write tests). 

It works great except that I don't think there is a way to check if written values are reasonable. I would love to be able to return 0x03 (Illegal Data Value) when someone tries to write an invalid data value. 

I would love to contribute this functionality (if you are open for it), but I am not sure how to go about it. 

Maybe something like this?
```go
package modbus

import (
	"github.com/simpleiot/simpleiot/modbus"
)

func main() {
	port := ...

	regs := modbus.Regs{}
	rtuTransport := modbus.NewRTU(port)
	srv := modbus.NewServer(modbusID, rtuTransport, &regs, 0)

	regs.AddReg(1, 1)
	regs.AddRegValueValidator(1, validateValue) // add a validator for register 1 (called when this value is written)

	srv.Listen(...)
}

// check if between 5 and 10
func validateValue(value uint16) bool {
	return value >= 5 && value <= 10
}
```

**Comment by cbrake** [comment]
that looks good to me!  Using a function makes it very flexible, so I think your proposal is good!

I've also considered moving the modbus module lib to a separate repo/package (github.com/simpleiot/modbus). There are advantages either way.

Curious, are you using the respreader as well, or how are you doing framing?

I've been considering adding parsing for framing (like https://github.com/simonvetter/modbus does) instead of timing.

**Comment by floriaanpost** [comment]
Yes, I do use respreader currently. Do I understand correctly that it buffers data until no new data has been received for some period (chunkTimeout) and then forwards it to the wrapped reader? 

As I understand, [https://github.com/simonvetter/modbus](https://github.com/simonvetter/modbus) is checking the size as it is streaming in, and know when the frame complete that way. I have done something similar in the past for another project and it does sound more efficient.

I could try to play around with that a bit and see how hard it would be to do it that way..

**Comment by cbrake** [comment]
> Yes, I do use respreader currently. Do I understand correctly that it buffers data until no new data has been received for some period (chunkTimeout) and then forwards it to the wrapped reader?

Yes, that is correct.
 
> As I understand, https://github.com/simonvetter/modbus is checking the size as it is streaming in, and know when the frame complete that way. I have done something similar in the past for another project and it does sound more efficient.
> 
> I could try to play around with that a bit and see how hard it would be to do it that way..

I think it would be a good option to have if it can be done simply and reliably. 

I used the respreader as I was already using it for other non-modbus protocols, so it was a shortcut to getting things done, but the code is tricky.

**Comment by cbrake** [comment]
Thinking about this a bit more, one common use case for validation is when a remote modbus devices is writing values into coils/holding regs we are controlling. In this case, we want to return the relevant exception codes (see section 7 in the modbus spec): 

https://modbus.org/docs/Modbus_Application_Protocol_V1_1b.pdf

so 03 `ILLEGAL_DATA_VALUE` should be returned if a write fails your the validator test. I don't think we are doing this right now -- been awhile since I wrote this code, so may be forgetting.
